docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,793 @@
+a quota management systems signals you have exhausted all the resources assigned to you.
+
+A stop-gap solution to this problem has been introduced in Cluster API with the implementation of machine deletion
+hooks, but this approach is tightly linked to the specific use case and it can not be re-used in other contexts/in
+other lifecycle moments.
+
+This proposal aims to solve the above problem in a more structured and generic way, by introducing the Runtime SDK,
+a set of rules, recommendations and fundamental capabilities required to implement a new extensibility mechanism
+that will allow systems, products and services built on top of Cluster API to interact with the Workload Cluster’s
+lifecycle. The key elements of this extensibility mechanism are Runtime Hooks and Runtime Extensions.
+
+### Goals
+
+To define the Runtime SDK and more specifically
+
+- To define the rules ensuring Runtime Hooks can evolve over time:
+  - When/how to modify the current version;
+  - When/how to create a new version;
+  - When/how to deprecate an old version, as well as mechanisms to inform users about versions being deprecated;
+  - When/how to drop an old version, as well as providing a mechanism to prevent users to upgrade Cluster API when
+    this will break installed Runtime Extensions;
+- To define the fundamental capabilities/tooling to be implemented in CAPI in order to allow the implementation of
+  Runtime Hooks.
+- To provide an initial set of guidelines for Runtime Extension developers.
+- To define how external Runtime Extensions can be registered within the Cluster API Runtime.
+
+### Non-Goals
+
+- To identify or specify the list of Runtime Hooks that should be implemented; some examples of possible Runtime Hooks
+  will be eventually provided, but it is not in the scope of this document to define them in detail;
+- To replace controllers or any other component of the Cluster API Runtime (including infrastructure providers,
+  bootstrap providers, control plane providers and the CRD/contract based extension mechanism they rely on).
+- To support the implementation of Runtime Hooks anywhere else outside the main Cluster API repository.
+
+### Future Work
+
+- Identify and specify the list of Runtime Hooks to be implemented; this will be addressed iteratively by a set of
+  future proposals, all of them building on top of the foundational capabilities introduced by this document;
+- Eventually consider deprecation of machine deletion hooks and replacement with a Runtime Hook;
+- Improve the Runtime Extension developer guide based on experience and feedback;
+- Add metrics about Runtime Extension calls (usage, usage vs deprecated versions, duration, error rate etc.);
+- Allow providers to use the same SDK to define their own hooks.
+
+## Proposal
+
+### User Stories
+
+- As a cluster operator I want to be able to execute a particular action in well-defined moments of the Workload
```

Comment:
Can we give a "particular action" concrete example so is immediately clear?

Reply:
+1 on concrete examples (even multiple would be nice)

Reply:
added a couple of examples. we can eventually make this list grow.

Reply:
@enxebre @CecileRobertMichon 
added a couple of examples; while doing this I also added a use case for externally generated patches, which is another potential use case for this SDK even if it does not qualify as a lifecycle event; I have tried to qualify this in a couple of point of the proposal (lifecycle hooks are the main goal, other usage are _by opportunity_)